### ISSUE TITLE
Expose corner radius on Rectangle shape

### ISSUE THREAD

**Issue by MarchingCube** [AvaloniaUI/Avalonia#4066]
When porting themes from UWP I found out that it is using `X` and `Y` corner radius on a `Rectangle.`
https://docs.microsoft.com/en-us/dotnet/api/system.windows.shapes.rectangle.radiusx?view=netcore-3.1#System_Windows_Shapes_Rectangle_RadiusX

It is not exposed on `RectangleGeometry` though (I guess since they don't want to check for rounded corners everywhere when clipping).

We could introduce `RoundedRectangleGeometry` and then make `Rectangle` make a decision when creating underlying geometry. Or make `RectangleGeometry` deal with corner radius as well.

On rendering side we are already covered since `RectangleNode` already supports rounded corners.

Any suggestions about API design on this?

**Comment by danwalmsley** [comment]
@MarchingCube I came across this, and every time replaced it with a border, however there is no reason we couldnt support this in future.

btw if you notice any other things, or leave anything out the port please make a note under:

https://github.com/AvaloniaUI/Avalonia/issues/4039

**Comment by kekekeks** [comment]
We need rectangle geometry with individual corner radii for setting up a geometry clip on Border as well.

**Comment by Gillibald** [comment]
BorderRenderHelper has everything to calculate a geometry for non uniform border with non uniform corner radii.

**Comment by kekekeks** [comment]
Note that Skia has specialized methods for SKRRect that work way more efficiently than ones that have to deal with SKPath. So it has to be implemented on the backend level.

**Comment by MarchingCube** [comment]
In Skia we use `SKPath` for all `Geometry` derived shapes. I think they do their own path simplification but it might be worth to benchmark it.

**Comment by jhofinger** [comment]
Note that contrary to UWP, WPF exposes RadiusX and RadiusY on RectangleGeometry (https://docs.microsoft.com/dotnet/api/system.windows.media.rectanglegeometry?view=netcore-3.1).

**Comment by robloo** [comment]
I'm going to pick this one up by the weekend.

I had a quick look and @Gillibald is correct. All the code to create the necessary geometry exists already for Border in BorderRenderHelper. That can be used internally within Rectangle to create the geometry, Rectangle only needs two new properties for RadiusX and RadiusY.

The question is should RectangleGeometry be modified at the same time?

**Comment by robloo** [comment]
I added this to `Rectangle` and it seems to work well enough to close out this issue. I decided against adding anything to `RectangleGeometry` as it really should be tied with changes to `IPlatformRenderInterface`. The Rendering interface itself needs to be able to handle corner radius on rectangles to keep things efficient. That should be a new issue and new PR in my opinion. If the core team disagrees, feel free to keep this open after the PR for Rectangle.
